File: sdks/python/src/wigolo/_local.py
"""Embedded local-daemon management.

``ensure_local_daemon`` probes a local port and, if nothing usable is
listening, spawns a local wigolo server and waits for it to become
healthy. It returns a ``LocalDaemon`` handle whose ``close()`` stops the
daemon only if this process owns it.

Zero runtime dependencies: stdlib only.
"""

from __future__ import annotations

import atexit
import collections
import json
import os
import shutil
import subprocess
import sys
import threading
import time
import urllib.error
import urllib.request
from typing import TYPE_CHECKING, Any, Optional

from ._errors import WigoloError

if TYPE_CHECKING:
    from ._client import Client

__all__ = ["LocalDaemon", "ensure_local_daemon", "local_client"]

# A dedicated opener that ignores http_proxy/https_proxy/no_proxy env: loopback
# traffic must NEVER route through a proxy (it breaks local mode and would leak
# the bearer token through the proxy). An empty ProxyHandler disables proxies.
_NO_PROXY_OPENER = urllib.request.build_opener(urllib.request.ProxyHandler({}))

_DEFAULT_LOCAL_PORT = 3333
_HEALTH_PROBE_TIMEOUT = 1.0
_SPAWN_HEALTH_BUDGET_S = 20.0
_SPAWN_POLL_INTERVAL_S = 0.25
_STDERR_RING_LINES = 20
_CLOSE_TERM_WAIT_S = 5.0


class _HealthResult:
    __slots__ = ("status_code",)

    def __init__(self, status_code: Optional[int]) -> None:
        self.status_code = status_code


def _probe_health(base_url: str, timeout: float = _HEALTH_PROBE_TIMEOUT) -> _HealthResult:
    """GET /health. Returns the HTTP status code, or None if refused/unreachable."""
    req = urllib.request.Request(f"{base_url}/health", method="GET")
    try:
        with _NO_PROXY_OPENER.open(req, timeout=timeout) as resp:
            return _HealthResult(resp.getcode())
    except urllib.error.HTTPError as exc:
        return _HealthResult(exc.code)
    except (urllib.error.URLError, ConnectionError, OSError):
        return _HealthResult(None)


def _probe_tools(base_url: str, token: Optional[str], timeout: float = _HEALTH_PROBE_TIMEOUT) -> _HealthResult:
    """GET /v1/tools with the resolved token. Returns the HTTP status code."""
    headers = {}
    if token:
        headers["Authorization"] = f"Bearer {token}"
    req = urllib.request.Request(f"{base_url}/v1/tools", method="GET", headers=headers)
    try:
        with _NO_PROXY_OPENER.open(req, timeout=timeout) as resp:
            resp.read()
            return _HealthResult(resp.getcode())
    except urllib.error.HTTPError as exc:
        return _HealthResult(exc.code)
    except (urllib.error.URLError, ConnectionError, OSError):
        return _HealthResult(None)


def _resolve_command() -> list[str]:
    """Resolve the argv list to launch a wigolo server.

    Precedence: env ``WIGOLO_CLI`` (JSON list, or a single whole-string
    executable path) > ``shutil.which('wigolo')``. The returned list is the
    executable prefix WITHOUT the ``serve ...`` args appended.
    """
    raw = os.environ.get("WIGOLO_CLI")
    if raw:
        stripped = raw.strip()
        if stripped.startswith("["):
            try:
                parsed = json.loads(stripped)
            except json.JSONDecodeError as exc:
                raise WigoloError(
                    f"WIGOLO_CLI looks like a JSON list but did not parse: {exc}"
                ) from exc
            if not isinstance(parsed, list) or not all(isinstance(x, str) for x in parsed):
                raise WigoloError("WIGOLO_CLI JSON must be a list of strings")
            return parsed
        # Not JSON: the WHOLE string is one executable path (may contain spaces).
        return [raw]

    which = shutil.which("wigolo")
    if which:
        return [which]

    raise WigoloError(
        "wigolo CLI not found — install a REST-capable wigolo (this SDK needs the "
        "REST API, newer than 0.1.43-beta.2) or set WIGOLO_CLI to the executable "
        "path (or a JSON argv list)."
    )


def _validate_port(value: Any, source: str) -> int:
    """Coerce/validate a port to an int in 1–65535, else raise WigoloError.

    A string must be a strict integer (no trailing junk, no float); an int/
    numeric value is range-checked. Actionable message names the ``source``.
    """
    if isinstance(value, bool):  # bool is an int subclass — reject it explicitly.
        raise WigoloError(f"{source} is not a valid port ({value!r}) — use 1–65535.")
    if isinstance(value, int):
        n = value
    elif isinstance(value, str):
        s = value.strip()
        if not s.isdigit():
            raise WigoloError(
                f"{source} is not a valid port ({value!r}) — set it to an integer "
                f"between 1 and 65535."
            )
        n = int(s)
    else:
        raise WigoloError(
            f"{source} is not a valid port ({value!r}) — set it to an integer "
            f"between 1 and 65535."
        )
    if not 1 <= n <= 65535:
        raise WigoloError(
            f"{source} is out of range ({n}) — set it to an integer between 1 "
            f"and 65535."
        )
    return n


def _redact_token(text: str, token: Optional[str]) -> str:
    """Replace every occurrence of the resolved bearer token with a marker."""
    if not token:
        return text
    return text.replace(token, "[redacted]")


def _is_windows() -> bool:
    return sys.platform == "win32"


def _needs_cmd_wrapper(exe: str) -> bool:
    if not _is_windows():
        return False
    lower = exe.lower()
    return lower.endswith((".cmd", ".bat", ".ps1"))


class LocalDaemon:
    """Handle to a local wigolo daemon (spawned or reused)."""

    def __init__(
        self,
        base_url: str,
        *,
        owned: bool,
        process: Optional[subprocess.Popen] = None,
        stderr_ring: Optional[collections.deque] = None,
    ) -> None:
        self.base_url = base_url
        self.owned = owned
        self._process = process
        self._stderr_ring = stderr_ring
        self._closed = False
        if owned and process is not None:
            atexit.register(self._atexit_close)

    def _atexit_close(self) -> None:
        try:
            self.close()
        except Exception:
            pass

    def stderr_tail(self) -> str:
        if self._stderr_ring is None:
            return ""
        return "".join(self._stderr_ring)

    def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        if not self.owned or self._process is None:
            return
        proc = self._process
        if proc.poll() is not None:
            return
        if _is_windows():
            try:
                subprocess.run(
                    ["taskkill", "/PID", str(proc.pid), "/T", "/F"],
                    check=False,
                    capture_output=True,
                )
            except Exception:
                pass
            try:
                proc.wait(timeout=_CLOSE_TERM_WAIT_S)
            except Exception:
                pass
            return
        # POSIX: escalate terminate -> kill. Mandatory: server shutdown awaits
        # open connections, so an in-flight long call would otherwise hang.
        try:
            proc.terminate()
        except ProcessLookupError:
            return
        try:
            proc.wait(timeout=_CLOSE_TERM_WAIT_S)
            return
        except subprocess.TimeoutExpired:
            pass
        try:
            proc.kill()
        except ProcessLookupError:
            return
        try:
            proc.wait(timeout=_CLOSE_TERM_WAIT_S)
        except Exception:
            pass


def _start_stderr_reader(proc: subprocess.Popen) -> collections.deque:
    ring: collections.deque = collections.deque(maxlen=_STDERR_RING_LINES)

    def _reader() -> None:
        assert proc.stderr is not None
        try:
            for line in proc.stderr:
                if isinstance(line, bytes):
                    line = line.decode("utf-8", errors="replace")
                ring.append(line)
        except Exception:
            pass

    t = threading.Thread(target=_reader, daemon=True)
    t.start()
    return ring


def _spawn(command: list[str], port: int, token: Optional[str]) -> subprocess.Popen:
    argv = list(command) + ["serve", "--port", str(port), "--host", "127.0.0.1"]

    child_env = dict(os.environ)
    child_env.pop("WIGOLO_DAEMON_HOST", None)
    child_env.pop("WIGOLO_DAEMON_PORT", None)
    if token:
        child_env["WIGOLO_API_TOKEN"] = token
    else:
        child_env.pop("WIGOLO_API_TOKEN", None)

    popen_kwargs: dict[str, Any] = {
        "stdout": subprocess.DEVNULL,
        "stderr": subprocess.PIPE,
        "env": child_env,
        "text": True,
    }

    if _is_windows():
        exe = command[0]
        if _needs_cmd_wrapper(exe):
            # CreateProcess can't exec .cmd/.bat/.ps1 directly.
            argv = ["cmd", "/c"] + argv
        creationflags = 0
        creationflags |= getattr(subprocess, "CREATE_NO_WINDOW", 0)
        popen_kwargs["creationflags"] = creationflags
    else:
        popen_kwargs["shell"] = False

    try:
        return subprocess.Popen(argv, **popen_kwargs)
    except (FileNotFoundError, PermissionError, OSError) as exc:
        raise WigoloError(
            f"wigolo CLI not launchable ({exc}) — install a REST-capable wigolo "
            "(this SDK needs the REST API, newer than 0.1.43-beta.2) or set "
            "WIGOLO_CLI to the executable path (or a JSON argv list)."
        ) from exc


def _wait_for_health(daemon_proc: subprocess.Popen, base_url: str, ring: collections.deque) -> bool:
    """Poll /health until 200 within the spawn budget. Returns True on healthy.

    Returns False (caller decides) if the child exits early or the budget
    elapses.
    """
    deadline = time.monotonic() + _SPAWN_HEALTH_BUDGET_S
    while time.monotonic() < deadline:
        if daemon_proc.poll() is not None:
            return False
        result = _probe_health(base_url)
        if result.status_code == 200:
            return True
        time.sleep(_SPAWN_POLL_INTERVAL_S)
    return False


def ensure_local_daemon(
    *,
    token: Optional[str],
    base_url_override: Optional[str] = None,
    port: Optional[int] = None,
    command: Optional[list[str]] = None,
) -> LocalDaemon:
    """Probe-or-spawn a local wigolo daemon.

    In local mode ``base_url_override`` (from ``WIGOLO_BASE_URL`` or an
    explicit ``base_url`` arg) is IGNORED — the daemon lives at
    ``http://127.0.0.1:{port}``.
    """
    if port is not None:
        resolved_port = _validate_port(port, "port argument")
    else:
        env_port = os.environ.get("WIGOLO_LOCAL_PORT")
        resolved_port = (
            _validate_port(env_port, "WIGOLO_LOCAL_PORT")
            if env_port
            else _DEFAULT_LOCAL_PORT
        )
    base_url = f"http://127.0.0.1:{resolved_port}"

    # Step 2: probe existing daemon.
    health = _probe_health(base_url)
    if health.status_code == 200:
        return _adopt_or_reject(base_url, resolved_port, token)
    # 503-down is NOT reusable; any non-200 that still answered means something
    # is listening but not a healthy reusable daemon — but per contract only a
    # refused connection triggers spawn. A 503 health means a daemon exists but
    # is unhealthy: fall through to reuse-check via tools probe is NOT allowed
    # (503 not reusable). We treat non-200-but-answered as spawn-blocking only
    # when the port is truly refused. Here status_code is not None => something
    # answered health but not 200 -> not reusable, and spawning would collide.
    if health.status_code is not None:
        raise WigoloError(
            f"a server on port {resolved_port} reports unhealthy (/health "
            f"{health.status_code}) and is not reusable — wait for it to recover "
            f"or set WIGOLO_LOCAL_PORT to another port."
        )

    # Step 3: connection refused -> spawn.
    cmd = command if command is not None else _resolve_command()
    proc = _spawn(cmd, resolved_port, token)
    ring = _start_stderr_reader(proc)

    healthy = _wait_for_health(proc, base_url, ring)
    if healthy:
        # Our child came up healthy (_wait_for_health returns False if the child
        # exited first, so a spawn-race loser is already handled below). We own
        # a daemon WE spawned with a matching token, so no capability probe is
        # needed here — the reuse path (_adopt_or_reject) is the only place that
        # probes /v1/tools, because that is where token/version mismatch matters.
        return LocalDaemon(base_url, owned=True, process=proc, stderr_ring=ring)

    # Not healthy: child may have early-exited (possibly bind error because a
    # winner raced us) or the budget elapsed.
    if proc.poll() is not None:
        # Child exited. Re-probe: a racing winner may be healthy.
        reprobe = _probe_health(base_url)
        if reprobe.status_code == 200:
            # Downgrade to reuse. Our child is already reaped (poll() != None).
            return _adopt_or_reject(base_url, resolved_port, token)
        tail = _redact_token("".join(ring), token)
        raise WigoloError(
            f"local wigolo daemon exited before becoming healthy on port "
            f"{resolved_port}. Try running `wigolo serve` manually to see the "
            f"error.\n--- server stderr (last {_STDERR_RING_LINES} lines) ---\n{tail}"
        )

    # Budget elapsed with child still running but never healthy: kill + raise.
    tail = _redact_token("".join(ring), token)
    daemon = LocalDaemon(base_url, owned=True, process=proc, stderr_ring=ring)
    daemon.close()
    raise WigoloError(
        f"local wigolo daemon did not become healthy within "
        f"{int(_SPAWN_HEALTH_BUDGET_S)}s on port {resolved_port}. Try running "
        f"`wigolo serve` manually to see the error.\n"
        f"--- server stderr (last {_STDERR_RING_LINES} lines) ---\n{tail}"
    )


def _adopt_or_reject(base_url: str, port: int, token: Optional[str]) -> LocalDaemon:
    """Given a health-200 daemon, decide whether it is a reusable wigolo.

    Capability probe GET /v1/tools with the resolved token:
      200 -> reuse (owned=False)
      401 -> never adopt (token mismatch / required)
      404 -> predates the REST API
      other/unparseable -> non-wigolo shape
    """
    tools = _probe_tools(base_url, token)
    code = tools.status_code
    if code == 200:
        return LocalDaemon(base_url, owned=False)
    if code == 401:
        raise WigoloError(
            f"a daemon on port {port} requires a bearer token this client doesn't "
            f"have (or the token mismatches) — set WIGOLO_API_TOKEN to match or pick "
            f"another WIGOLO_LOCAL_PORT."
        )
    if code == 404:
        raise WigoloError(
            f"a daemon on port {port} predates the REST API — upgrade the server or "
            f"set WIGOLO_LOCAL_PORT to another port."
        )
    raise WigoloError(
        f"the server on port {port} does not look like a REST-capable wigolo daemon "
        f"(/v1/tools returned {code}) — set WIGOLO_LOCAL_PORT to another port."
    )


def local_client(**opts: Any) -> "Client":
    """Construct a ``Client`` in embedded local mode.

    Accepts the same options as ``Client`` (base_url is ignored in local
    mode). Convenience wrapper for ``Client(local=True, ...)``.
    """
    from ._client import Client

    opts.pop("local", None)
    return Client(local=True, **opts)

File: sdks/typescript/src/local.ts
/**
 * Embedded local-daemon mode (node-only). `createLocalClient` ensures a wigolo
 * daemon is listening on a loopback port — reusing an already-running one, or
 * spawning `wigolo serve` when the port is free — then hands back a
 * {@link WigoloClient} plus lifecycle controls. This is the zero-setup path a
 * connection-refused error points at.
 *
 * Exposed at the "wigolo-sdk/local" subpath. It imports `node:*` builtins, so
 * it is intentionally NOT part of the edge-safe barrel.
 */
import { spawn, type ChildProcess } from 'node:child_process';
import { existsSync } from 'node:fs';
import { delimiter, join } from 'node:path';
import { WigoloClient, type WigoloClientOptions } from './client.js';
import { WigoloError } from './errors.js';

const LOOPBACK_HOST = '127.0.0.1';
const DEFAULT_PORT = 3333;
const HEALTH_PROBE_TIMEOUT_MS = 1000;
const SPAWN_HEALTH_BUDGET_MS = 20_000;
const HEALTH_POLL_INTERVAL_MS = 250;
const CLOSE_SIGTERM_GRACE_MS = 5_000;
const STDERR_RING_LINES = 20;

/** The version below which the daemon lacks the REST API this SDK requires. */
const MIN_REST_VERSION_HINT = '0.1.43-beta.2';

export interface CreateLocalClientOptions {
  /** Loopback port. Default resolution: option > WIGOLO_LOCAL_PORT > 3333. */
  port?: number;
  /**
   * Command (argv array, end-to-end — never split on spaces) used to spawn the
   * daemon. Default resolution: option > WIGOLO_CLI (JSON array, or a single
   * executable path) > `wigolo` resolved on PATH.
   */
  command?: string[];
  /** Bearer token. Default resolution: option > WIGOLO_API_TOKEN. */
  token?: string;
  /** Per-request deadline override handed to the returned client. */
  timeoutMs?: number;
}

export interface LocalClient {
  /** The client bound to the local daemon. */
  client: WigoloClient;
  /** True when this call spawned the daemon (and `close()` will stop it). */
  owned: boolean;
  /** Stop the owned daemon (no-op when not owned). */
  close(): Promise<void>;
}

function readEnv(name: string): string | undefined {
  try {
    const value = process.env[name];
    return typeof value === 'string' && value.length > 0 ? value : undefined;
  } catch {
    return undefined;
  }
}

/**
 * Validate a port as an integer in 1–65535. Accepts a number (option) or the
 * raw env string; rejects trailing garbage like "3333;x" that parseInt would
 * silently truncate to 3333.
 */
function validatePort(value: number | string, source: string): number {
  let n: number;
  if (typeof value === 'number') {
    n = value;
  } else {
    // A strict integer string only — no trailing junk, no floats.
    if (!/^\d+$/.test(value.trim())) {
      throw new WigoloError(
        `${source} is not a valid port ("${value}") — set it to an integer between 1 and 65535.`,
      );
    }
    n = Number.parseInt(value.trim(), 10);
  }
  if (!Number.isInteger(n) || n < 1 || n > 65535) {
    throw new WigoloError(
      `${source} is out of range (${n}) — set it to an integer between 1 and 65535.`,
    );
  }
  return n;
}

function resolvePort(opts: CreateLocalClientOptions): number {
  if (opts.port !== undefined) return validatePort(opts.port, 'port option');
  const envPort = readEnv('WIGOLO_LOCAL_PORT');
  if (envPort) return validatePort(envPort, 'WIGOLO_LOCAL_PORT');
  return DEFAULT_PORT;
}

const isWindows = process.platform === 'win32';

/** Resolve an executable name against PATH (PATHEXT-aware on win32). */
function resolveOnPath(name: string): string | null {
  if (name.includes('/') || (isWindows && name.includes('\\'))) {
    return existsSync(name) ? name : null;
  }
  const pathValue = process.env.PATH ?? '';
  const dirs = pathValue.split(delimiter).filter(Boolean);
  const exts = isWindows ? (process.env.PATHEXT ?? '.COM;.EXE;.BAT;.CMD').split(';') : [''];
  for (const dir of dirs) {
    for (const ext of exts) {
      const candidate = join(dir, name + ext);
      if (existsSync(candidate)) return candidate;
    }
  }
  return null;
}

/**
 * Resolve the base spawn command (argv, sans `serve` args). Never split on
 * spaces: a WIGOLO_CLI string that is not a JSON array is treated as ONE
 * executable path (paths can contain spaces).
 */
function resolveCommand(opts: CreateLocalClientOptions): string[] {
  if (opts.command && opts.command.length > 0) return opts.command;
  const envCli = readEnv('WIGOLO_CLI');
  if (envCli) {
    const trimmed = envCli.trim();
    if (trimmed.startsWith('[')) {
      try {
        const parsed = JSON.parse(trimmed) as unknown;
        if (Array.isArray(parsed) && parsed.every((x) => typeof x === 'string') && parsed.length > 0) {
          return parsed as string[];
        }
      } catch {
        // Fall through: treat the whole string as one path.
      }
    }
    return [envCli];
  }
  const resolved = resolveOnPath('wigolo');
  if (resolved) return [resolved];
  // Return the bare name; spawn will surface an actionable ENOENT.
  return ['wigolo'];
}

/** Whether a resolved command ends in a Windows shim extension needing cmd. */
function needsCmdShim(command: string): boolean {
  const lower = command.toLowerCase();
  return lower.endsWith('.cmd') || lower.endsWith('.bat') || lower.endsWith('.ps1');
}

interface ProbeResult {
  reachable: boolean;
  healthStatus?: number;
}

async function probeHealth(base: string, token: string | undefined): Promise<ProbeResult> {
  try {
    const res = await fetch(`${base}/health`, {
      method: 'GET',
      headers: token ? { Authorization: `Bearer ${token}` } : {},
      signal: AbortSignal.timeout(HEALTH_PROBE_TIMEOUT_MS),
    });
    return { reachable: true, healthStatus: res.status };
  } catch {
    return { reachable: false };
  }
}

/**
 * Capability probe against a reachable daemon. GET /v1/tools with the client's
 * token: 200 → reusable; anything else maps to an actionable throw.
 */
async function probeCapability(base: string, port: number, token: string | undefined): Promise<void> {
  let res: Response;
  try {
    res = await fetch(`${base}/v1/tools`, {
      method: 'GET',
      headers: token ? { Authorization: `Bearer ${token}` } : {},
      signal: AbortSignal.timeout(HEALTH_PROBE_TIMEOUT_MS * 3),
    });
  } catch (err) {
    throw new Error(
      `A daemon is answering on port ${port} but its tool index could not be read (${String(err)}). ` +
        `Pick another WIGOLO_LOCAL_PORT or restart the daemon.`,
    );
  }
  if (res.status === 200) return;
  if (res.status === 401 || res.status === 403) {
    throw new Error(
      `The daemon on port ${port} requires a bearer token this client doesn't have (or the token doesn't match). ` +
        `Set WIGOLO_API_TOKEN to match the running daemon, or pick another WIGOLO_LOCAL_PORT.`,
    );
  }
  if (res.status === 404) {
    throw new Error(
      `The daemon on port ${port} predates the REST API this SDK needs. ` +
        `Upgrade the server (newer than ${MIN_REST_VERSION_HINT}) or set WIGOLO_LOCAL_PORT to a REST-capable daemon.`,
    );
  }
  throw new Error(
    `The service on port ${port} does not look like a REST-capable wigolo daemon (GET /v1/tools returned ${res.status}). ` +
      `Set WIGOLO_LOCAL_PORT to a different port.`,
  );
}

/** Bounded ring buffer of the most recent stderr lines from the child. */
class StderrRing {
  private lines: string[] = [];
  push(chunk: string): void {
    for (const line of chunk.split(/\r?\n/)) {
      if (line.length === 0) continue;
      this.lines.push(line);
      if (this.lines.length > STDERR_RING_LINES) this.lines.shift();
    }
  }
  text(): string {
    return this.lines.join('\n');
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

const TASKKILL_WAIT_MS = 2_000;

/** Best-effort tree/plain kill of a child, platform-aware. */
function killChild(child: ChildProcess, signal: NodeJS.Signals): void {
  if (child.pid === undefined || child.killed) return;
  if (isWindows) {
    try {
      spawn('taskkill', ['/PID', String(child.pid), '/T', '/F'], { windowsHide: true });
    } catch {
      /* best effort */
    }
    return;
  }
  try {
    child.kill(signal);
  } catch {
    /* best effort */
  }
}

/**
 * win32-only: spawn taskkill and await its exit (bounded) so a caller's
 * `close()` does not resolve while the tree-kill is still in flight.
 */
async function killChildWindowsAwait(child: ChildProcess): Promise<void> {
  if (child.pid === undefined || child.killed) return;
  await new Promise<void>((resolve) => {
    let done = false;
    const finish = (): void => {
      if (done) return;
      done = true;
      resolve();
    };
    let killer: ChildProcess;
    try {
      killer = spawn('taskkill', ['/PID', String(child.pid), '/T', '/F'], { windowsHide: true });
    } catch {
      finish();
      return;
    }
    const timer = setTimeout(finish, TASKKILL_WAIT_MS);
    killer.on('exit', () => {
      clearTimeout(timer);
      finish();
    });
    killer.on('error', () => {
      clearTimeout(timer);
      finish();
    });
  });
}

export async function createLocalClient(opts: CreateLocalClientOptions = {}): Promise<LocalClient> {
  const port = resolvePort(opts);
  const base = `http://${LOOPBACK_HOST}:${port}`;
  const token = opts.token !== undefined ? opts.token : readEnv('WIGOLO_API_TOKEN');

  const clientOptions: WigoloClientOptions = { baseUrl: base };
  if (token !== undefined) clientOptions.token = token;
  if (opts.timeoutMs !== undefined) clientOptions.timeoutMs = opts.timeoutMs;
  const makeClient = (): WigoloClient => new WigoloClient(clientOptions);

  // Step 2: probe an existing daemon.
  const probe = await probeHealth(base, token);
  if (probe.reachable) {
    // 200 required for reuse; a 503-down daemon is not reusable.
    if (probe.healthStatus === 200) {
      await probeCapability(base, port, token);
      return { client: makeClient(), owned: false, close: async () => {} };
    }
    throw new Error(
      `A daemon on port ${port} reports unhealthy (HTTP ${probe.healthStatus}). ` +
        `Wait for it to recover, restart it, or set WIGOLO_LOCAL_PORT to another port.`,
    );
  }

  // Step 3: connection refused → spawn our own.
  return spawnLocalDaemon({ base, port, token, makeClient, opts });
}

interface SpawnArgs {
  base: string;
  port: number;
  token: string | undefined;
  makeClient: () => WigoloClient;
  opts: CreateLocalClientOptions;
}

async function spawnLocalDaemon(args: SpawnArgs): Promise<LocalClient> {
  const { base, port, token, makeClient, opts } = args;
  const baseCommand = resolveCommand(opts);
  const serveArgs = ['serve', '--port', String(port), '--host', LOOPBACK_HOST];

  // Child env: inherit minus daemon host/port overrides; force the token to
  // match this client so daemon and client always agree.
  const childEnv: NodeJS.ProcessEnv = { ...process.env };
  delete childEnv.WIGOLO_DAEMON_HOST;
  delete childEnv.WIGOLO_DAEMON_PORT;
  if (token !== undefined) childEnv.WIGOLO_API_TOKEN = token;
  else delete childEnv.WIGOLO_API_TOKEN;

  const executable = baseCommand[0];
  let spawnCmd: string;
  let spawnArgs: string[];
  if (isWindows && needsCmdShim(executable)) {
    // Direct .cmd spawn is EINVAL on patched Node (CVE-2024-27980) — go via cmd.
    spawnCmd = 'cmd';
    spawnArgs = ['/c', executable, ...baseCommand.slice(1), ...serveArgs];
  } else {
    spawnCmd = executable;
    spawnArgs = [...baseCommand.slice(1), ...serveArgs];
  }

  const ring = new StderrRing();
  let child: ChildProcess;
  try {
    child = spawn(spawnCmd, spawnArgs, {
      env: childEnv,
      stdio: ['ignore', 'ignore', 'pipe'],
      windowsHide: true,
    });
  } catch (err) {
    throw spawnLaunchError(err);
  }

  child.stderr?.on('data', (buf: Buffer) => ring.push(buf.toString('utf-8')));

  let exited = false;
  let exitInfo: { code: number | null; signal: NodeJS.Signals | null } | null = null;
  child.on('exit', (code, signal) => {
    exited = true;
    exitInfo = { code, signal };
  });
  let spawnErr: unknown;
  child.on('error', (err) => {
    spawnErr = err;
    exited = true;
  });

  const exitKillHook = (): void => killChild(child, 'SIGKILL');
  process.on('exit', exitKillHook);

  const owned: LocalClient = {
    client: makeClient(),
    owned: true,
    close: async () => {
      process.removeListener('exit', exitKillHook);
      if (exited) return;
      if (isWindows) {
        await killChildWindowsAwait(child);
        return;
      }
      killChild(child, 'SIGTERM');
      const deadline = Date.now() + CLOSE_SIGTERM_GRACE_MS;
      while (!exited && Date.now() < deadline) {
        await sleep(50);
      }
      if (!exited) killChild(child, 'SIGKILL');
    },
  };

  // Step 4: poll /health until 200 within budget.
  const deadline = Date.now() + SPAWN_HEALTH_BUDGET_MS;
  while (Date.now() < deadline) {
    if (exited) {
      // Step 5: concurrent-spawn race — a rival may have won the port bind.
      const rival = await probeHealth(base, token);
      if (rival.reachable && rival.healthStatus === 200) {
        process.removeListener('exit', exitKillHook);
        await probeCapability(base, port, token);
        return { client: makeClient(), owned: false, close: async () => {} };
      }
      process.removeListener('exit', exitKillHook);
      if (spawnErr) throw spawnLaunchError(spawnErr);
      throw childExitError(port, exitInfo, ring, token);
    }
    const health = await probeHealth(base, token);
    if (health.reachable && health.healthStatus === 200) {
      return owned;
    }
    await sleep(HEALTH_POLL_INTERVAL_MS);
  }

  // Timed out waiting for health — kill and report.
  process.removeListener('exit', exitKillHook);
  killChild(child, 'SIGKILL');
  throw new Error(
    `Local daemon on port ${port} did not become healthy within ${SPAWN_HEALTH_BUDGET_MS}ms. ` +
      `Try running \`wigolo serve --port ${port}\` manually to see the full output.` +
      stderrTail(ring, token),
  );
}

function spawnLaunchError(err: unknown): Error {
  const code = (err as { code?: string }).code;
  if (code === 'ENOENT' || code === 'EINVAL' || code === 'EACCES') {
    return new Error(
      `The wigolo CLI could not be launched (${code}). ` +
        `Install a REST-capable wigolo (this SDK needs the REST API, newer than ${MIN_REST_VERSION_HINT}), ` +
        `or set WIGOLO_CLI to the executable (or a JSON argv array).`,
    );
  }
  return new Error(`Failed to launch the wigolo CLI: ${String(err)}`);
}

function childExitError(
  port: number,
  exitInfo: { code: number | null; signal: NodeJS.Signals | null } | null,
  ring: StderrRing,
  token: string | undefined,
): Error {
  const how = exitInfo?.signal
    ? `signal ${exitInfo.signal}`
    : `exit code ${exitInfo?.code ?? 'unknown'}`;
  return new Error(
    `The local daemon exited before becoming healthy on port ${port} (${how}). ` +
      `Try running \`wigolo serve --port ${port}\` manually to see the full output.` +
      stderrTail(ring, token),
  );
}

/** Redact every occurrence of the resolved bearer token from a string. */
function redactToken(text: string, token: string | undefined): string {
  if (!token) return text;
  return text.split(token).join('[redacted]');
}

function stderrTail(ring: StderrRing, token: string | undefined): string {
  const tail = redactToken(ring.text(), token);
  return tail.length > 0 ? `\n--- daemon stderr (last lines) ---\n${tail}` : '';
}
